git: change extraConfig from str to attrs

uvNikita opened the pull request:
This makes it possible to configure git like this:
```
programs.git {
  enable = true;
  extraConfig = {
    pull = {
      rebase = true;
    };
  };
};
```

Instead of specifying the string like this:
```
programs.git {
  enable = true;
  extraConfig = ''
    [pull]
    rebase=true
  '';
};
```

Unfortunately, it doesn't allow to define sections in multiple modules, e.g. if it will be another module which will have `programs.git.extraConfig.pull = { ff = true; }`, then the final configuration will only contain one of the options:
```
[pull]
ff=1
```
At the same time, string value doesn't allow to extend individual sections either.

For now couldn't find the reason to use string value instead of attributes, correct me if I'm wrong.

rycee commented:
I don't know why I didn't use an attribute set here to start with. I imagine it having to do with the point you make with options being overwritten. In general it shouldn't be too much of a problem. But I still think it is quite severe, in particular it is impossible to, e.g., say
```
programs.git.signing.key = "blahblah";
programs.git.extraConfig.commit.verbose = true;`
```
because `programs.git.signing.key` will write to the `commit` section of the INI file.

I think it should be possible to have `extraConfig` as an attribute set and still have the above case work by changing your PR as follows:

```patch
diff --git a/modules/programs/git.nix b/modules/programs/git.nix
index 88e0dd4..f1f6bb5 100644
--- a/modules/programs/git.nix
+++ b/modules/programs/git.nix
@@ -69,30 +69,43 @@ in
         default = {};
         description = "Additional configuration to add.";
       };
+
+      iniContent = mkOption {
+        internal = true;
+        type = types.attrsOf types.attrs;
+      };
     };
   };
 
   config = mkIf cfg.enable (
-    let
-      ini = {
-        user = {
+    mkMerge [
+      {
+        home.packages = [ cfg.package ];
+
+        programs.git.iniContent.user = {
           name = cfg.userName;
           email = cfg.userEmail;
-        } // optionalAttrs (cfg.signing != null) {
-          signingKey = cfg.signing.key;
         };
-      } // optionalAttrs (cfg.signing != null) {
-        commit.gpgSign = cfg.signing.signByDefault;
-        gpg.program = cfg.signing.gpgPath;
-      } // optionalAttrs (cfg.aliases != {}) {
-        alias = cfg.aliases;
-      } // cfg.extraConfig;
-    in
-      {
-        home.packages = [ cfg.package ];
 
         home.file.".gitconfig".text =
-          generators.toINI {} ini;
+          generators.toINI {} cfg.iniContent;
+      }
+
+      (mkIf (cfg.signing != null) {
+        programs.git.iniContent = {
+          user.signingKey = cfg.signing.key;
+          commit.gpgSign = cfg.signing.signByDefault;
+          gpg.program = cfg.signing.gpgPath;
+        };
+      })
+
+      (mkIf (cfg.aliases != {}) {
+        programs.git.iniContent.alias = cfg.aliases;
+      })
+
+      {
+        programs.git.iniContent = cfg.extraConfig;
       }
+    ]
   );
 }
```

If this is OK then the problem of backwards compatibility still remains. Unfortunately it wouldn't work very well to have `condition = isString config.programs.git.extraConfig` or something because it will error out during evaluation. What I can imagine is to now make a news entry, conditioned on `extraConfig` being non-empty, that say that the option in the near future will change type…

uvNikita commented:
I changed the code according to your diff, and now it works as it should, thank you for the suggestion! As for backward compatibility, I managed to support both types using `types.either`. So now both old and new configuration styles are supported until the sting type option will be removed. I also added news entry with the corresponding condition.

rycee commented on modules/misc/news.nix:
```diff
@@ -192,6 +192,19 @@ in
+        time = "2017-09-22T12:09:01+00:00";
+        condition = isString config.programs.git.extraConfig;
+        message = ''
+          The 'config.programs.git.extraConfig' parameter is now
```
I think this should be just "The 'programs.git.extraConfig' parameter is now …". I.e., without the "config." part.

rycee commented on modules/misc/news.nix:
```diff
@@ -192,6 +192,19 @@ in
+          The 'config.programs.git.extraConfig' parameter is now
+          accepting attributes instead of strings which allows more
+          flexible configuration.
+
```
Maybe nice to add an example here. Something like

```
          For example, if your configuration includes

              programs.git.extraConfig = '''
                [core]
                editor = vim
              ''';

          then you can now change it to

              programs.git.extraConfig = {
                core = {
                  editor = "vim";
                };
              };
```

perhaps?

rycee commented:
Very nice! This looks good and works well for me (both with string and attrs).

Maybe change the commit message to "git: change extraConfig from lines to attrs"? Since it actually is of type lines :-)

uvNikita commented:
I agree with suggestions, changed news entry and commit message.

rycee commented:
Good work! Rebased to master in 0f096f9ad40f2fb055f4b696f16450d7487cd895. Many thanks :-)
